libs/ledger-live-common/src/apps/inlineAppInstall.ts

```diff
@@ -0,0 +1,122 @@
+import Transport from "@ledgerhq/hw-transport";
+import { Observable, concat, of, from, EMPTY, defer } from "rxjs";
+import { ConnectAppEvent } from "../hw/connectApp";
+import getDeviceInfo from "../hw/getDeviceInfo";
+import { listApps, execWithTransport } from "./hw";
+import {
+  reducer,
+  initState,
+  isOutOfMemoryState,
+  predictOptimisticState,
+} from "./logic";
+import { runAllWithProgress } from "./runner";
+import { InlineAppInstallEvent } from "./types";
+import { mergeMap, map } from "rxjs/operators";
+
+/**
+ * Tries to install a list of apps
+ *
+ * @param transport Transport instance
+ * @param appNames List of app names to install
+ * @param onSuccessObs Optional observable to run after the installation
+ * @param allowPartialDependencies If true, keep installing apps even if some are missing
+ * @returns Observable of InlineAppInstallEvent or ConnectAppEvent
+ * - Event "inline-install" contains a global progress of the installation
+ */
+const inlineAppInstall = ({
+  transport,
+  appNames,
+  onSuccessObs,
+  allowPartialDependencies = false,
+}: {
+  transport: Transport;
+  appNames: string[];
+  onSuccessObs?: () => Observable<any>;
+  allowPartialDependencies?: boolean;
+}): Observable<InlineAppInstallEvent | ConnectAppEvent> =>
+  concat(
+    of({
+      type: "listing-apps",
+    }),
+    from(getDeviceInfo(transport)).pipe(
+      mergeMap((deviceInfo) => listApps(transport, deviceInfo)),
+      mergeMap((e) => {
+        // Bubble up events
+        if (
+          e.type === "device-permission-granted" ||
+          e.type === "device-permission-requested"
+        ) {
+          return of(e);
+        }
+
+        if (e.type === "result") {
+          // Figure out the operations that need to happen
+          const state = appNames.reduce(
+            (state, name) =>
+              reducer(state, {
+                type: "install",
+                name,
+                allowPartialDependencies,
+              }),
+            initState(e.result)
+          );
+
+          // Failed appOps in this flow will throw by default but if we're here
+          // it means we didn't throw, so we wan't to notify the action about it.
+          const maybeSkippedEvent: Observable<InlineAppInstallEvent> = state
```

Comment:
Why would we have skipped event here ? We did not try to install any apps yet, right ?